commit a62d90f2e5cb55d0c8998b63120d6db3c55df866
Author: Thomas Munro <tmunro@postgresql.org>
Date:   Wed Nov 27 22:56:41 2024 +1300

    Revert "Blind attempt to fix _configthreadlocale() failures on MinGW."
    
    This reverts commit 2cf91ccb73ce888c44e3751548fb7c77e87335f2.
    
    When using the old msvcrt.dll, MinGW would supply its own dummy version
    of _configthreadlocale() that just returns -1 if you try to use it.  For
    a time we tolerated that to shut the build farm up.  We would fall back
    to code that was enough for the tests to pass, but it would surely have
    risked crashing a real multithreaded program.
    
    We don't need that kludge anymore, because we can count on ucrt.  We
    expect the real _configthreadlocale() to be present, and the ECPG tests
    will now fail if it isn't.  The workaround was dead code and it's time
    to revert it.
    
    (A later patch still under review proposes to remove this use of
    _configthreadlocale() completely but we're unwinding this code in
    steps.)
    
    Reviewed-by: Peter Eisentraut <peter@eisentraut.org>
    Discussion: https://postgr.es/m/d9e7731c-ca1b-477c-9298-fa51e135574a%40eisentraut.org

diff --git a/src/interfaces/ecpg/ecpglib/descriptor.c b/src/interfaces/ecpg/ecpglib/descriptor.c
index 56e2bc41531..aee888432f0 100644
--- a/src/interfaces/ecpg/ecpglib/descriptor.c
+++ b/src/interfaces/ecpg/ecpglib/descriptor.c
@@ -510,11 +510,11 @@ ECPGget_desc(int lineno, const char *desc_name, int index,...)
 			setlocale(LC_NUMERIC, stmt.oldlocale);
 			ecpg_free(stmt.oldlocale);
 		}
 #ifdef WIN32
 		if (stmt.oldthreadlocale != -1)
-			(void) _configthreadlocale(stmt.oldthreadlocale);
+			_configthreadlocale(stmt.oldthreadlocale);
 #endif
 #endif
 	}
 	else if (data_var.ind_type != ECPGt_NO_INDICATOR && data_var.ind_pointer != NULL)
 
diff --git a/src/interfaces/ecpg/ecpglib/execute.c b/src/interfaces/ecpg/ecpglib/execute.c
index 466d5600f9b..b5089eac787 100644
--- a/src/interfaces/ecpg/ecpglib/execute.c
+++ b/src/interfaces/ecpg/ecpglib/execute.c
@@ -1975,13 +1975,11 @@ ecpg_do_prologue(int lineno, const int compat, const int force_indicator,
 
 	/*
 	 * Make sure we do NOT honor the locale for numeric input/output since the
 	 * database wants the standard decimal point.  If available, use
 	 * uselocale() for this because it's thread-safe.  Windows doesn't have
-	 * that, but it usually does have _configthreadlocale().  In some versions
-	 * of MinGW, _configthreadlocale() exists but always returns -1 --- so
-	 * treat that situation as if the function doesn't exist.
+	 * that, but it does have _configthreadlocale().
 	 */
 #ifdef HAVE_USELOCALE
 
 	/*
 	 * Since ecpg_init() succeeded, we have a connection.  Any successful
@@ -1995,10 +1993,15 @@ ecpg_do_prologue(int lineno, const int compat, const int force_indicator,
 		return false;
 	}
 #else
 #ifdef WIN32
 	stmt->oldthreadlocale = _configthreadlocale(_ENABLE_PER_THREAD_LOCALE);
+	if (stmt->oldthreadlocale == -1)
+	{
+		ecpg_do_epilogue(stmt);
+		return false;
+	}
 #endif
 	stmt->oldlocale = ecpg_strdup(setlocale(LC_NUMERIC, NULL), lineno);
 	if (stmt->oldlocale == NULL)
 	{
 		ecpg_do_epilogue(stmt);
@@ -2216,21 +2219,16 @@ ecpg_do_epilogue(struct statement *stmt)
 #ifdef HAVE_USELOCALE
 	if (stmt->oldlocale != (locale_t) 0)
 		uselocale(stmt->oldlocale);
 #else
 	if (stmt->oldlocale)
+	{
 		setlocale(LC_NUMERIC, stmt->oldlocale);
 #ifdef WIN32
-
-	/*
-	 * This is a bit trickier than it looks: if we failed partway through
-	 * statement initialization, oldthreadlocale could still be 0.  But that's
-	 * okay because a call with 0 is defined to be a no-op.
-	 */
-	if (stmt->oldthreadlocale != -1)
-		(void) _configthreadlocale(stmt->oldthreadlocale);
+		_configthreadlocale(stmt->oldthreadlocale);
 #endif
+	}
 #endif
 
 	free_statement(stmt);
 }
 
